File: products/sdlc-cc/platform/packages/insights-core/python/insights_core/publisher.py
"""Non-blocking signal publisher for llm-gateway/dlp/rag/etc."""

from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass
from typing import Protocol

from insights_core.types import SignalEvent


class Transport(Protocol):
    async def publish(self, subject: str, data: bytes) -> None: ...


@dataclass
class Stats:
    published: int = 0
    failed: int = 0
    dropped: int = 0


class Publisher:
    """Fire-and-forget publisher. Emit returns synchronously; actual send runs
    on a background task. Back-pressure drops events to protect the hot path.
    """

    def __init__(self, transport: Transport, *, queue_size: int = 1024) -> None:
        self._transport = transport
        self._queue: asyncio.Queue[tuple[str, bytes]] = asyncio.Queue(maxsize=queue_size)
        self._stats = Stats()
        self._task: asyncio.Task[None] | None = None

    async def start(self) -> None:
        if self._task is None or self._task.done():
            self._task = asyncio.create_task(self._run(), name="insights-publisher")

    async def stop(self) -> None:
        if self._task:
            await self._queue.put(("__stop__", b""))
            await self._task
            self._task = None

    def emit(self, ev: SignalEvent) -> None:
        if not ev.source or not ev.tenant_id:
            raise ValueError("publisher: source and tenant_id required")
        subject = f"signals.{ev.source.value}.{ev.event_type}"
        payload = json.dumps(ev.to_json()).encode()
        try:
            self._queue.put_nowait((subject, payload))
        except asyncio.QueueFull:
            self._stats.dropped += 1

    async def _run(self) -> None:
        while True:
            subject, data = await self._queue.get()
            if subject == "__stop__":
                return
            try:
                await self._transport.publish(subject, data)
                self._stats.published += 1
            except Exception:
                self._stats.failed += 1

    @property
    def stats(self) -> Stats:
        return Stats(
            published=self._stats.published,
            failed=self._stats.failed,
            dropped=self._stats.dropped,
        )

File: products/sdlc-cc/platform/packages/insights-core/go/publisher/publisher.go
package publisher

import (
	"context"
	"encoding/json"
	"errors"
	"sync/atomic"
	"time"

	"github.com/sdlc-ai/platform/packages/insights-core/types"
)

// Transport is what a JetStream client must satisfy. Decoupled so callers can
// swap in a real nats.JetStreamContext or a test fake.
type Transport interface {
	Publish(ctx context.Context, subject string, data []byte) error
}

// Publisher is non-blocking by default: Emit returns immediately after
// enqueueing; failures increment a counter but never block the hot path.
type Publisher struct {
	t         Transport
	queue     chan envelope
	dropped   atomic.Uint64
	published atomic.Uint64
	failed    atomic.Uint64
}

type envelope struct {
	subject string
	payload []byte
}

type Options struct {
	QueueSize int
}

func New(t Transport, opts Options) *Publisher {
	size := opts.QueueSize
	if size <= 0 {
		size = 1024
	}
	p := &Publisher{t: t, queue: make(chan envelope, size)}
	go p.run()
	return p
}

// Emit enqueues a SignalEvent for publication. Returns immediately. If the
// queue is full the event is dropped and counted; callers must accept loss
// under back-pressure — durability belongs to JetStream once accepted.
func (p *Publisher) Emit(ev types.SignalEvent) error {
	if ev.Source == "" || ev.TenantID == "" {
		return errors.New("publisher: source and tenant_id required")
	}
	b, err := json.Marshal(ev)
	if err != nil {
		return err
	}
	subject := "signals." + string(ev.Source) + "." + ev.EventType
	select {
	case p.queue <- envelope{subject: subject, payload: b}:
		return nil
	default:
		p.dropped.Add(1)
		return errors.New("publisher: queue full")
	}
}

func (p *Publisher) run() {
	for env := range p.queue {
		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		if err := p.t.Publish(ctx, env.subject, env.payload); err != nil {
			p.failed.Add(1)
		} else {
			p.published.Add(1)
		}
		cancel()
	}
}

// Stats exposes counters for Prometheus wiring by the host service.
type Stats struct {
	Published uint64
	Failed    uint64
	Dropped   uint64
}

func (p *Publisher) Stats() Stats {
	return Stats{
		Published: p.published.Load(),
		Failed:    p.failed.Load(),
		Dropped:   p.dropped.Load(),
	}
}

func (p *Publisher) Close() { close(p.queue) }
